ruby/sdk-phase-2.md

```diff
@@ -0,0 +1,575 @@
+    unwrap-from-future-fail in there. It can get confusing for users to catch too.
+* We chose not to make whether `any_of` raises be configurable. There is a lot of confusion about what return value
+  and exception situation should be across the industry in different languages. We're taking a very simple approach
+  and if users struggle with `any_of` and needing to know which future succeeded/failed, they can be more specific in
+  their return/raise inside their future block.
+  * We chose to add an `any_of_no_raise` that does not raise.
+    * 💭 Why not just an arg? Because it would change return value which is a confusing thing to do just based on arg.
+    * 💭 Why not `any_of!`? Because it's not necessarily more dangerous needing `!`, just a different behavior.
+* We chose not to make whether `all_of` raises be configurable. There is a lot of confusion about what return value
+  and exception situation should be across the industry in different languages. We're taking a very simple approach
+  and if users struggle with `all_of` need to wait for other futures to succeed when one fails, they need to make sure
+  none fail (can return exception instead or something).
+  * We chose to add an `all_of_no_raise` that does not raise.
+    * 💭 Why not just an arg? Because we have `any_of_no_raise` and this is consistent with that.
+    * 💭 Why not `all_of!`? Because it's not necessarily more dangerous needing `!`, just a different behavior.
+
+Example of running many activities and waiting on all to complete:
+
+```ruby
+class ManyActivityWorkflow < Temporalio::Workflow
+  def execute(count)
+    futures = count.times.map do
+      Future.new do
+        Workflow.execute_activity(MyActivity, 'some arg' start_to_close_timeout: 5)
+      end
+    end
+    Temporalio::Workflow::Future.all_of(*futures).wait
+  end
+end
+```
+
+### Task Running
+
+Temporal workflow tasks for the workflow worker portion are short-lived CPU-bound work. While these usually only take a
+few milliseconds to run, they still use a thread and so we need to give the user some control over this.
+
+* We will extract the thread pool from `Temporalio::Worker::ActivityExecutor::ThreadPool` to
+  `Temporalio::Worker::ThreadPool` (and have activity executor use it).
+* `Temporalio::Worker::WorkflowTaskExecutor` will exist with two implementations
+  * `Temporalio::Worker::WorkflowTaskExecutor::ThreadPool` that accepts a thread pool which defaults to max threads
+    unbounded.
+    * There is a `default` class method with a lazily created global default.
+    * ❓ Is this an acceptable default? Python uses `[Etc.nprocessors, 4].max` as the max threads, but we figure it can
+      be unbounded here and remain bounded by the Core `max_concurrent_workflow_tasks` (TODO: expose this, it was left
+      off in `main`).
+    * ❓ Should the default share the same thread pool as activity executor?
```

Comment:
I don't think so, specifically because of some of the workflow isolation stuff that may or may not need to happen in them. Seems easier and cleaner to keep them separate without much downside.

Reply:
There is a difference between thread pool and executor in this case. One benefit of reusing the thread pool (but still different executors of course) is that it is an "unbounded caching thread pool" which means the same threads could be reused across workflows and activities. The threads, once returned back to the pool, should have no side effects from its use. By having separate unbounded thread pools here, both pools will have idle threads that the other could be using unnecessarily. Definitely not an easy/hard thing to share vs keep separate because we will have lazily created global defaults for these (and already do for activity executor).